samza-core/src/main/java/org/apache/samza/operators/impl/WatermarkStates.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.operators.impl;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.system.WatermarkMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class manages the state of the watermarks in a task. Internally it keeps track of the latest
+ * watermark timestamp from each upstream task, and use the min as the consolidated watermark time.
```

Comment:
This is a bit confusing. The WatermarkState is per task, while inputWatermark/outputWatermark are local variables in each operatorImpl. So, can I assume that the logic is:
1) the first input operator impl invokes aggregateWatermark() method and records the current min watermark per ssp
2) based on the current min watermark per ssp, the onWatermark() method checks the corresponding operatorImpl's inputWatermark and tries to advance watermark if current min watermark per ssp > the operator's inputWatermark
3) the operatorImpl may generate outputWatermark and propagate it downstream, which is used by the downstream operatorImpl to update their inputWatermark and generate outputWatermark
So, at the end, when the watermarks are propagated at the boundary of a task, can we assume that the outputWatermark at the boundary is >= the shared WatermarkState per ssp in the task? I would suggest to add an example with numbers to explain the relationship between the WatermarkState here in the task, and the inputWatermark/outputWatermark as local variables in the OperatorImpl.